File: api/src/integration/java/org/pmiops/workbench/NotebooksIntegrationTest.java
package org.pmiops.workbench;

import static com.google.common.truth.Truth.assertThat;

import org.junit.jupiter.api.Test;
import org.pmiops.workbench.db.dao.WorkspaceDao;
import org.pmiops.workbench.firecloud.FireCloudService;
import org.pmiops.workbench.firecloud.FirecloudApiClientFactory;
import org.pmiops.workbench.google.StorageConfig;
import org.pmiops.workbench.leonardo.LeonardoApiClient;
import org.pmiops.workbench.leonardo.LeonardoApiClientFactory;
import org.pmiops.workbench.leonardo.LeonardoApiClientImpl;
import org.pmiops.workbench.utils.mappers.LeonardoMapper;
import org.pmiops.workbench.workspaces.WorkspaceAuthService;
import org.pmiops.workbench.workspaces.resources.UserRecentResourceService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Import;

public class NotebooksIntegrationTest extends BaseIntegrationTest {
  @Autowired private LeonardoApiClient leonardoApiClient;

  // Provide mock beans for dependencies of NotebooksServiceImpl (which is loaded as a bean within
  // this test due to the @ComponentScan on the o.p.w.notebooks package.
  @MockBean FireCloudService mockFireCloudService;
  @MockBean UserRecentResourceService mockUserRecentResourceService;
  @MockBean LeonardoMapper leonardoMapper;
  @MockBean WorkspaceDao workspaceDao;
  @MockBean WorkspaceAuthService workspaceAuthService;

  @TestConfiguration
  // N.B. in the other integration test classes we add a @ComponentScan which scans the package
  // where the class under test is defined. Adding that annotation
  @ComponentScan(basePackageClasses = LeonardoApiClientImpl.class)
  @Import({
    FirecloudApiClientFactory.class,
    LeonardoApiClientImpl.class,
    LeonardoApiClientFactory.class,
    StorageConfig.class,
    BaseIntegrationTest.Configuration.class
  })
  static class Configuration {}

  @Test
  public void testStatus() {
    assertThat(leonardoApiClient.getLeonardoStatus()).isTrue();
  }
}

File: api/src/integration/java/org/pmiops/workbench/NotebooksIntegrationTest.kt
package org.pmiops.workbench

import com.google.common.truth.Truth.assertThat

import org.junit.Test
import org.mockito.Mock
import org.pmiops.workbench.config.WorkbenchConfig
import org.pmiops.workbench.notebooks.LeonardoNotebooksClient
import org.pmiops.workbench.notebooks.LeonardoNotebooksClientImpl
import org.pmiops.workbench.notebooks.NotebooksRetryHandler
import org.pmiops.workbench.notebooks.api.ClusterApi
import org.pmiops.workbench.notebooks.api.NotebooksApi
import org.pmiops.workbench.test.Providers
import org.pmiops.workbench.workspaces.WorkspaceService
import org.springframework.retry.backoff.NoBackOffPolicy

/** Created by brubenst on 5/8/18.  */
class NotebooksIntegrationTest {
    /*
   * Mocked service providers are (currently) not available for functional integration tests.
   * Integration tests with real projects/clusters against production Notebooks is not
   * recommended at this time.
   */

    @Mock
    private val clusterApi: ClusterApi? = null
    @Mock
    private val notebooksApi: NotebooksApi? = null
    @Mock
    private val workspaceService: WorkspaceService? = null

    private val leonardoNotebooksClient = LeonardoNotebooksClientImpl(
            Providers.of<ClusterApi>(clusterApi),
            Providers.of<NotebooksApi>(notebooksApi),
            Providers.of(createConfig()),
            Providers.of<User>(null),
            NotebooksRetryHandler(NoBackOffPolicy()),
            workspaceService)

    @Test
    fun testStatus() {
        assertThat(leonardoNotebooksClient.notebooksStatus).isTrue()
    }

    private fun createConfig(): WorkbenchConfig {
        val config = WorkbenchConfig()
        config.firecloud = WorkbenchConfig.FireCloudConfig()
        config.firecloud.debugEndpoints = true
        return config
    }
}
